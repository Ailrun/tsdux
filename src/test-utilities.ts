import * as ts from 'typescript';
import * as setting from '../ts-tests/tsconfig.json';

const compileTs = (...fileNames: Array<string>): Array<string> => {
  const options = ts.convertCompilerOptionsFromJson(setting.compilerOptions, '.').options;
  const program = ts.createProgram(fileNames, options);

  const emitResult = program.emit();
  const allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
  const allMessages = allDiagnostics
    .map((diagnostic) => {
      const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');

      if (diagnostic.file && diagnostic.start) {
        const { fileName } = diagnostic.file;
        const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
        const formattedMessage = message.split('\n')
          .map(s => '    ' + s)
          .join('\n');

        return `${fileName}/${line + 1},${character + 1} :\n${formattedMessage}`;
      } else {
        return message;
      }
    });

  return allMessages;
}

export {
  compileTs,
};
