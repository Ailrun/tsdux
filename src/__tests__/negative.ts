import { compileTs as compile } from '../test-utilities';

test('action', () => {
  expect(compile('./fail-tests/action.ts')).toMatchSnapshot();
});

test('actiontype', () => {
  expect(compile('./fail-tests/actiontype.ts')).toMatchSnapshot();
});

test('reducer', () => {
  expect(compile('./fail-tests/reducer.ts')).toMatchSnapshot();
});
