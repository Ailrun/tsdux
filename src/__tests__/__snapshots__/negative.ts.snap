// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action 1`] = `
Array [
  "ts-tests/action.ts/41,8 : Argument of type '5' is not assignable to parameter of type 'string'.",
  "ts-tests/action.ts/42,8 : Argument of type 'true' is not assignable to parameter of type 'string'.",
  "ts-tests/action.ts/43,8 : Argument of type 'undefined[]' is not assignable to parameter of type 'string'.",
  "ts-tests/action.ts/44,8 : Argument of type '{}' is not assignable to parameter of type 'string'.",
  "ts-tests/action.ts/47,7 : Type 'ActionCreatorWithoutPayload<\\"action/FAILED\\">' is not assignable to type 'ActionCreatorWithPayload<\\"action/FAILED\\", {}>'.
  Types of property 'action' are incompatible.
    Type 'ActionWithoutPayload<\\"action/FAILED\\">' is not assignable to type 'ActionWithPayload<\\"action/FAILED\\", {}>'.
      Property 'payload' is missing in type 'ActionWithoutPayload<\\"action/FAILED\\">'.",
  "ts-tests/action.ts/48,7 : Type 'ActionCreatorWithoutPayload<\\"action/W/O/PL1\\">' is not assignable to type 'ActionCreatorWithPayload<\\"action/W/O/PL1\\", {}>'.
  Types of property 'action' are incompatible.
    Type 'ActionWithoutPayload<\\"action/W/O/PL1\\">' is not assignable to type 'ActionWithPayload<\\"action/W/O/PL1\\", {}>'.
      Property 'payload' is missing in type 'ActionWithoutPayload<\\"action/W/O/PL1\\">'.",
  "ts-tests/action.ts/51,8 : Argument of type '5' is not assignable to parameter of type 'string'.",
  "ts-tests/action.ts/52,8 : Argument of type 'RegExp' is not assignable to parameter of type 'string'.",
  "ts-tests/action.ts/53,8 : Argument of type 'undefined[]' is not assignable to parameter of type 'string'.",
  "ts-tests/action.ts/56,7 : Type 'ActionCreatorWithPayload<\\"action/FAILED\\", number>' is not assignable to type 'ActionCreatorWithoutPayload<\\"action/FAILED\\">'.
  Types of property 'create' are incompatible.
    Type '(payload: number) => ActionWithPayload<\\"action/FAILED\\", number>' is not assignable to type '() => ActionWithoutPayload<\\"action/FAILED\\">'.",
  "ts-tests/action.ts/57,7 : Type 'ActionCreatorWithPayload<\\"action/WITH_PL2\\", string>' is not assignable to type 'ActionCreatorWithoutPayload<\\"action/WITH_PL2\\">'.
  Types of property 'create' are incompatible.
    Type '(payload: string) => ActionWithPayload<\\"action/WITH_PL2\\", string>' is not assignable to type '() => ActionWithoutPayload<\\"action/WITH_PL2\\">'.",
  "ts-tests/action.ts/60,1 : Expected 0 arguments, but got 1.",
  "ts-tests/action.ts/61,1 : Expected 0 arguments, but got 2.",
  "ts-tests/action.ts/64,7 : Type 'ActionWithoutPayload<\\"action/W/O/PL1\\">' is not assignable to type 'ActionWithoutPayload<\\"IDK\\">'.
  Type '\\"action/W/O/PL1\\"' is not assignable to type '\\"IDK\\"'.",
  "ts-tests/action.ts/65,7 : Type 'ActionWithoutPayload<\\"action/W/O/PL0\\">' is not assignable to type 'ActionWithoutPayload<\\"GEWFEWK\\">'.
  Type '\\"action/W/O/PL0\\"' is not assignable to type '\\"GEWFEWK\\"'.",
  "ts-tests/action.ts/68,7 : Type 'ActionWithoutPayload<\\"action/W/O/PL1\\">' is not assignable to type 'ActionWithPayload<\\"action/W/O/PL2\\", {}>'.
  Property 'payload' is missing in type 'ActionWithoutPayload<\\"action/W/O/PL1\\">'.",
  "ts-tests/action.ts/69,7 : Type 'ActionWithoutPayload<\\"action/W/O/PL1\\">' is not assignable to type 'ActionWithPayload<\\"action/W/O/PL1\\", {}>'.",
  "ts-tests/action.ts/72,1 : Expected 1 arguments, but got 0.",
  "ts-tests/action.ts/73,1 : Expected 1 arguments, but got 2.",
  "ts-tests/action.ts/74,1 : Expected 1 arguments, but got 3.",
  "ts-tests/action.ts/80,7 : Type 'ActionWithPayload<\\"action/WITH_PL0\\", {}>' is not assignable to type 'ActionWithPayload<\\"gwer\\", {}>'.
  Type '\\"action/WITH_PL0\\"' is not assignable to type '\\"gwer\\"'.",
  "ts-tests/action.ts/81,7 : Type 'ActionWithPayload<\\"action/WITH_PL3\\", string[]>' is not assignable to type 'ActionWithPayload<\\"aciton/WITH_PL2\\", string[]>'.
  Type '\\"action/WITH_PL3\\"' is not assignable to type '\\"aciton/WITH_PL2\\"'.",
  "ts-tests/action.ts/84,7 : Type 'ActionWithPayload<\\"action/WITH_PL1\\", number>' is not assignable to type 'ActionWithPayload<\\"action/WITH_PL1\\", string>'.
  Type 'number' is not assignable to type 'string'.",
  "ts-tests/action.ts/85,7 : Type 'ActionWithPayload<\\"action/WITH_PL0\\", {}>' is not assignable to type 'ActionWithPayload<\\"action/WITH_PL0\\", { value: number; }>'.
  Type '{}' is not assignable to type '{ value: number; }'.
    Property 'value' is missing in type '{}'.",
  "ts-tests/action.ts/93,20 : Cannot assign to 'payload' because it is a constant or a read-only property.",
  "ts-tests/action.ts/94,20 : Cannot assign to 'payload' because it is a constant or a read-only property.",
]
`;

exports[`actiontype 1`] = `
Array [
  "ts-tests/actiontype.ts/47,1 : Expected 1 arguments, but got 0.",
  "ts-tests/actiontype.ts/50,7 : Argument of type '12' is not assignable to parameter of type 'ActionCreator<string, any>[]'.",
  "ts-tests/actiontype.ts/51,7 : Argument of type '\\"fweweglu\\"' is not assignable to parameter of type 'ActionCreator<string, any>[]'.",
  "ts-tests/actiontype.ts/52,7 : Argument of type '{}' is not assignable to parameter of type 'ActionCreator<string, any>[]'.
  Property 'length' is missing in type '{}'.",
  "ts-tests/actiontype.ts/53,7 : Argument of type 'ActionCreatorWithPayload<\\"app/tests/ADD_TEST\\", Test>' is not assignable to parameter of type 'ActionCreator<string, any>[]'.
  Property 'length' is missing in type 'ActionCreatorWithPayload<\\"app/tests/ADD_TEST\\", Test>'.",
  "ts-tests/actiontype.ts/56,1 : Expected 1 arguments, but got 2.",
  "ts-tests/actiontype.ts/57,1 : Expected 1 arguments, but got 2.",
  "ts-tests/actiontype.ts/58,1 : Expected 1 arguments, but got 2.",
  "ts-tests/actiontype.ts/61,7 : Argument of type '(ActionCreatorWithPayload<\\"app/tests/ADD_TEST\\", Test> | ActionCreatorWithPayload<\\"app/tests/UPDAT...' is not assignable to parameter of type 'ActionCreator<string, any>[]'.
  Type 'ActionCreatorWithPayload<\\"app/tests/ADD_TEST\\", Test> | ActionCreatorWithPayload<\\"app/tests/UPDATE...' is not assignable to type 'ActionCreator<string, any>'.
    Type 'undefined[]' is not assignable to type 'ActionCreator<string, any>'.
      Type 'undefined[]' is not assignable to type 'ActionCreatorWithPayload<string, any>'.
        Property 'type' is missing in type 'undefined[]'.",
  "ts-tests/actiontype.ts/62,7 : Argument of type '(string | ActionCreatorWithPayload<\\"app/tests/ADD_TEST\\", Test> | ActionCreatorWithPayload<\\"app/te...' is not assignable to parameter of type 'ActionCreator<string, any>[]'.
  Type 'string | ActionCreatorWithPayload<\\"app/tests/ADD_TEST\\", Test> | ActionCreatorWithPayload<\\"app/tes...' is not assignable to type 'ActionCreator<string, any>'.
    Type 'string' is not assignable to type 'ActionCreator<string, any>'.",
  "ts-tests/actiontype.ts/65,7 : Type 'ActionWithPayload<\\"app/tests/ADD_TEST\\", Test> | ActionWithPayload<\\"app/tests/REMOVE_TEST\\", number>' is not assignable to type 'ActionWithPayload<\\"app/tests/ADD_TEST\\", Test>'.
  Type 'ActionWithPayload<\\"app/tests/REMOVE_TEST\\", number>' is not assignable to type 'ActionWithPayload<\\"app/tests/ADD_TEST\\", Test>'.
    Type '\\"app/tests/REMOVE_TEST\\"' is not assignable to type '\\"app/tests/ADD_TEST\\"'.",
  "ts-tests/actiontype.ts/66,7 : Type 'ActionWithPayload<\\"app/tests/ADD_TEST\\", Test> | ActionWithPayload<\\"app/tests/REMOVE_TEST\\", number>' is not assignable to type 'ActionWithPayload<\\"app/tests/REMOVE_TEST\\", number>'.
  Type 'ActionWithPayload<\\"app/tests/ADD_TEST\\", Test>' is not assignable to type 'ActionWithPayload<\\"app/tests/REMOVE_TEST\\", number>'.
    Type '\\"app/tests/ADD_TEST\\"' is not assignable to type '\\"app/tests/REMOVE_TEST\\"'.",
  "ts-tests/actiontype.ts/75,27 : Property 'toFixed' does not exist on type 'Test'.",
  "ts-tests/actiontype.ts/81,27 : Property 'expected' does not exist on type 'number'.",
  "ts-tests/actiontype.ts/88,19 : Property 'payload' does not exist on type 'ActionWithoutPayload<\\"app/tests/CHECK_TEST\\">'.",
  "ts-tests/actiontype.ts/93,17 : Property 'payload' does not exist on type 'ActionWithPayload<\\"app/tests/ADD_TEST\\", Test> | ActionWithPayload<\\"app/tests/UPDATE_TEST\\", { id: ...'.
  Property 'payload' does not exist on type 'ActionWithoutPayload<\\"app/tests/CHECK_TEST\\">'.",
]
`;

exports[`reducer 1`] = `Array []`;
